src/ring_theory/hahn_series.lean

```diff
@@ -382,14 +442,135 @@ noncomputable instance [comm_semiring R] : comm_semiring (hahn_series Γ R) :=
   end,
   .. hahn_series.semiring }
 
-noncomputable instance [ring R] : ring (hahn_series Γ R) :=
+instance [ring R] : ring (hahn_series Γ R) :=
 { .. hahn_series.semiring,
   .. hahn_series.add_comm_group }
 
-noncomputable instance [comm_ring R] : comm_ring (hahn_series Γ R) :=
+instance [comm_ring R] : comm_ring (hahn_series Γ R) :=
 { .. hahn_series.comm_semiring,
   .. hahn_series.ring }
 
+section semiring
+variables [semiring R]
+
+@[simp]
+lemma single_mul_single {a b : Γ} {r s : R} :
+  single a r * single b s = single (a + b) (r * s) :=
+begin
+  ext x,
+  by_cases h : x = a + b,
+  { rw [h, mul_single_coeff_add],
+    simp },
+  { rw [single_coeff_of_ne h, mul_coeff, finset.sum_eq_zero],
+    rintros ⟨y1, y2⟩ hy,
+    obtain ⟨rfl, hy1, hy2⟩ := finset.mem_add_antidiagonal.1 hy,
+    rw [eq_of_mem_support_single hy1, eq_of_mem_support_single hy2] at h,
+    exact (h rfl).elim }
+end
+
+/-- `C a` is the constant Hahn Series `a`. `C` is provided as a ring homomorphism. -/
+def C : R →+* (hahn_series Γ R) :=
+{ to_fun := single 0,
+  map_zero' := single_eq_zero,
+  map_one' := rfl,
+  map_add' := λ x y, by { ext a, by_cases h : a = 0; simp [h] },
+  map_mul' := λ x y, by rw [single_mul_single, zero_add] }
+
+@[simp]
+lemma C_apply (r : R) : C r = single (0 : Γ) r := rfl
+
+@[simp]
+lemma C_zero : C (0 : R) = (0 : hahn_series Γ R) := C.map_zero
+
+@[simp]
+lemma C_one : C (1 : R) = (1 : hahn_series Γ R) := C.map_one
+
+lemma C_mul_eq_smul {r : R} {x : hahn_series Γ R} : C r * x = r • x :=
+single_zero_mul_eq_smul
+
+end semiring
+
+section algebra
+variables [comm_semiring R] {A : Type*} [semiring A] [algebra R A]
+
+instance : algebra R (hahn_series Γ A) :=
+{ to_ring_hom := C.comp (algebra_map R A),
+  smul_def' := λ r x, by { ext, simp },
+  commutes' := λ r x, by { ext, simp only [smul_coeff, single_zero_mul_eq_smul, ring_hom.coe_comp,
+    ring_hom.to_fun_eq_coe, C_apply, function.comp_app, algebra_map_smul, mul_single_zero_coeff],
+    rw [← algebra.commutes, algebra.smul_def], }, }
+
+theorem C_eq_algebra_map : C = (algebra_map R (hahn_series Γ R)) := rfl
+
+theorem algebra_map_apply {r : R} :
+  algebra_map R (hahn_series Γ A) r = C (algebra_map R A r) := rfl
+
+end algebra
```

Comment:
Does it sound a reasonable statement to add? I think it is interesting, and worth notin (and absent from `power_series`.) Of course, I could provide the proof but first wanted to check if you agree.

Reply:
This comment is relative to commit `debaf7a` which I was stupid enough to push _after_ posting the comment.

Reply:
The statement you added (`instance : nontrivial (subalgebra R (hahn_series Γ R))`) needs `[nontrivial Γ] [nontrivial R]` and probably needs me to work out the theory of the constant subalgebra. I'd like to do that at some point, but is that necessary for this PR?

Reply:
I added it for `polynomial`, `power_series`, `mv_power_series`, and `hahn_series`.

Reply:
I think we already have a few things on general nontrivial subalgebra, but that they're missing `instance` annotations (#6025), but I might be misremembering